stablehlo/reference/Ops.cpp

```diff
@@ -800,6 +820,37 @@ Tensor evalClzOp(const Tensor &operand, ShapedType resultType) {
   return result;
 }
 
+Tensor evalCollectivePermuteOp(
+    const Tensor &operand, SmallVector<SmallVector<uint32_t>> sourceTargetPairs,
+    ChannelId channelId, Process *process) {
+  if (!process)
+    llvm::report_fatal_error(
+        "collective_permute is only supported when run via "
+        "interpreter.run_parallel");
+
+  ProcessGroups processGroups;
+  if (channelId <= 0) processGroups = process->crossReplica(sourceTargetPairs);
+  if (channelId > 0) processGroups = process->crossPartition(sourceTargetPairs);
+
+  std::optional<Tensor> result;
```

Comment:
Given that Tensor supports the notion of empty tensors, we don't need to represent optional tensors via `std::optional`. Just `Tensor` would suffice. The only thing missing is this:

```C++
class Tensor {
  ...
  explicit operator bool() const { return (bool)impl_; }
  bool operator!() const { return !impl_; }
  ...
}
```